engine/access/state_stream/backend/backend_events.go

```diff
@@ -43,27 +48,62 @@ func (b EventsBackend) SubscribeEvents(ctx context.Context, startBlockID flow.Id
 	return sub
 }
 
+// getResponseFactory returns a function function that returns the event response for a given height.
 func (b EventsBackend) getResponseFactory(filter state_stream.EventFilter) GetDataByHeightFunc {
-	return func(ctx context.Context, height uint64) (interface{}, error) {
-		executionData, err := b.getExecutionData(ctx, height)
-		if err != nil {
-			return nil, fmt.Errorf("could not get execution data for block %d: %w", height, err)
+	return func(ctx context.Context, height uint64) (response interface{}, err error) {
+		if b.useIndex {
+			response, err = b.getEventsFromStorage(height, filter)
+		} else {
+			response, err = b.getEventsFromExecutionData(ctx, height, filter)
 		}
 
-		events := []flow.Event{}
-		for _, chunkExecutionData := range executionData.ChunkExecutionDatas {
-			events = append(events, filter.Filter(chunkExecutionData.Events)...)
+		if err == nil && b.log.GetLevel() == zerolog.TraceLevel {
+			eventsResponse := response.(*EventsResponse)
+			b.log.Trace().
+				Hex("block_id", logging.ID(eventsResponse.BlockID)).
+				Uint64("height", height).
+				Int("events", len(eventsResponse.Events)).
+				Msg("sending events")
 		}
+		return
+	}
+}
+
+// getEventsFromExecutionData returns the events for a given height extractd from the execution data.
+func (b EventsBackend) getEventsFromExecutionData(ctx context.Context, height uint64, filter state_stream.EventFilter) (*EventsResponse, error) {
+	executionData, err := b.getExecutionData(ctx, height)
+	if err != nil {
+		return nil, fmt.Errorf("could not get execution data for block %d: %w", height, err)
+	}
+
+	var events flow.EventsList
+	for _, chunkExecutionData := range executionData.ChunkExecutionDatas {
+		events = append(events, filter.Filter(chunkExecutionData.Events)...)
+	}
 
-		b.log.Trace().
-			Hex("block_id", logging.ID(executionData.BlockID)).
-			Uint64("height", height).
-			Msgf("sending %d events", len(events))
+	return &EventsResponse{
+		BlockID: executionData.BlockID,
+		Height:  height,
+		Events:  events,
+	}, nil
+}
 
-		return &EventsResponse{
-			BlockID: executionData.BlockID,
-			Height:  height,
-			Events:  events,
-		}, nil
+// getEventsFromStorage returns the events for a given height from the index storage.
+func (b EventsBackend) getEventsFromStorage(height uint64, filter state_stream.EventFilter) (*EventsResponse, error) {
+	header, err := b.headers.ByHeight(height)
```

Comment:
```suggestion
	blockID, err := b.headers.BlockIDByHeight(height)
```

Reply:
applied in [522556f](https://github.com/onflow/flow-go/pull/5160/commits/522556f3df134429f7c4a382361066a513d12e88)